packages/app/src/composables/useCloudSpecData.ts

```diff
@@ -0,0 +1,119 @@
+import { gql, useMutation } from '@urql/vue'
+import { Ref, computed, watch } from 'vue'
+import { useDebounce } from '@vueuse/core'
+import { CloudData_RefetchDocument, SpecsListFragment } from '../generated/graphql'
+
+gql`
+mutation CloudData_Refetch ($ids: [ID!]!) {
+  loadRemoteFetchables(ids: $ids){
+    id
+    fetchingStatus
+  }
+}
+`
+
+type NonNullCloudSpec = Exclude<SpecsListFragment['cloudSpec'], undefined | null>
+
+/**
+ * Set up watchers & subscriptions to clear/refetch cloud data based on project state relative to latest data in the cloud
+ * @param isProjectDisconnected Whether project has a valid connection to the cloud
+ * @param isOffline  Whether app is currently offline (network state)
+ * @param projectId Cloud project ID (slug)
+ * @param mostRecentUpdate Date/time of the latest project update in the cloud
+ * @param displayedSpecs Spec entries currently displayed in the app
+ * @param allSpecs All project spec entries
+ * @returns Trigger functions
+ */
+export function useCloudSpecData (
+  isProjectDisconnected: Ref<boolean>,
+  isOffline: Ref<boolean>,
+  projectId: string | null | undefined,
+  mostRecentUpdate: Ref<string | null>,
+  displayedSpecs: Ref<(SpecsListFragment | undefined)[]>,
+  allSpecs: (SpecsListFragment | undefined)[],
+) {
+  const refetchMutation = useMutation(CloudData_RefetchDocument)
+
+  const isCloudSpecOlderThan = (item: NonNullCloudSpec, comparisonDttm: string | null) => {
+    if (item.data?.__typename !== 'CloudProjectSpec' && item.data?.__typename !== 'CloudProjectSpecNotFound') {
+      return false
+    }
+
+    if (!item.data.retrievedAt || !comparisonDttm) {
+      return false
+    }
+
+    return new Date(comparisonDttm).getTime() > new Date(item.data.retrievedAt).getTime()
+  }
+
+  const shouldRefetch = (item: NonNullCloudSpec) => {
+    if (isOffline.value) {
+      // Offline, no need to refetch
+      return false
+    }
+
+    if (item.fetchingStatus === 'NOT_FETCHED' || item.fetchingStatus === undefined) {
+      // Not yet fetched
+      return true
+    }
+
+    if (isCloudSpecOlderThan(item, mostRecentUpdate.value)) {
+      // outdated
+      return true
+    }
+
+    // nothing new, no need to refetch
+    return false
+  }
+
+  /**
+   * Refetch any displayed RemoteFetchable entries that are older than the latest cloudProject update
+   * or that have not yet been fetched
+   */
+  const fetchDisplayedCloudData = async () => {
+    const cloudSpecIdsToRefetch = displayedSpecs.value
```

Comment:
Should we combine these into a single `reduce` so we don't need to loop so much? Small optimization but always good to reduce iteration (imo).

Reply:
I agree this would minimize iteration, but I feel like the `reduce` would be more difficult to grok. Since the array length here should be <100-ish (only displayed rows) I think increased readability outweighs the potential perf gains. Also (and I may be overthinking this) would the `.map/.filter` approach be more efficient for building an array than repeated `.push` calls like the reduce would call for?

```
const cloudSpecIdsToRefetch = displayedSpecs.value
    .reduce((values, spec) => {
      if (spec?.cloudSpec?.id && shouldRefetch(spec.cloudSpec)) {
        values.push(spec.cloudSpec.id)
      }

      return values
    }, [] as string[])
```

Reply:
Fair enough... whenever I see `map` and `filter` combos my intuition is to merge them. To your point; we should just benchmark these things and find out exactly which one is faster and by how much. I'm guessing it's negligent in this case.

:+1:

Reply:
I looked into a lot of these when I was spiking into specs list perf more generally. I agree `reduce` is less readable at a glance, it's fine but I generally find other ways of achieving the same thing clearer. 

I do think `map -> filter -> map` situations can be optimized to drop the first map, so here we could examine the `cloudSpec` property directly in the `filter` instead of doing an iteration to extract that one property, another to examine it. 

That said I did some pretty unscientific logging to figure out where time was being spent in dealing with the specs list and this stuff was not definitely not the bottleneck. I refactored a few instances of "extra" loops and it didn't make any difference, even in larger lists.